donkeycar/parts/pytorch/torch_data.py

```diff
@@ -0,0 +1,149 @@
+# PyTorch
+import torch
+from torch.utils.data import IterableDataset, DataLoader
+from donkeycar.utils import train_test_split
+from donkeycar.parts.tub_v2 import Tub
+from torchvision import transforms
+from typing import List, Any
+from donkeycar.pipeline.types import TubRecord, TubDataset
+from donkeycar.pipeline.sequence import TubSequence
+import pytorch_lightning as pl
+
+def get_default_transform(for_video=False, for_inference=False):
+    """
+    Creates a default transform to work with torchvision models
+
+    Video transform:
+    All pre-trained models expect input images normalized in the same way, 
+    i.e. mini-batches of 3-channel RGB videos of shape (3 x T x H x W), 
+    where H and W are expected to be 112, and T is a number of video frames 
+    in a clip. The images have to be loaded in to a range of [0, 1] and 
+    then normalized using mean = [0.43216, 0.394666, 0.37645] and 
+    std = [0.22803, 0.22145, 0.216989].
+    """
+
+    mean = [0.485, 0.456, 0.406]
+    std = [0.229, 0.224, 0.225]
+    input_size = (224, 224)
+
+    if for_video:
+        mean = [0.43216, 0.394666, 0.37645]
+        std = [0.22803, 0.22145, 0.216989]
+        input_size = (112, 112)
+
+    transform = transforms.Compose([
+        transforms.Resize(input_size),
+        transforms.ToTensor(),
+        transforms.Normalize(mean=mean, std=std)
+    ])
+
+    return transform
+
+
+class TorchTubDataset(IterableDataset):
+    '''
+    Loads the dataset, and creates a train/test split.
+    '''
+
+    def __init__(self, config, records: List[TubRecord], transform=None):
+        """Create a PyTorch Tub Dataset
+
+        Args:
+            config (object): the configuration information
+            records (List[TubRecord]): a list of tub records
+            transform (function, optional): a transform to apply to the data
+        """
+        self.config = config
+
+        # Handle the transforms
+        if transform:
+            self.transform = transform
+        else:
+            self.transform = get_default_transform()
+
+        self.sequence = TubSequence(records)
+        self.pipeline = self._create_pipeline()
+
+    def _create_pipeline(self):
+        """ This can be overridden if more complicated pipelines are
+            required """
+
+        def y_transform(record: TubRecord):
+            angle: float = record.underlying['user/angle']
+            throttle: float = record.underlying['user/throttle']
+            return torch.tensor([angle, throttle])
+
+        def x_transform(record: TubRecord):
+            # Loads the result of Image.open()
+            img_arr = record.image(cached=True, as_nparray=False)
```

Comment:
Should I make `x_transform` and `y_transform` instance methods? Or should I leave as is?

Reply:
You should leave this as-is. Look at the comment above. Having a class here, gives us more flexibility to take advantage of what the ML framework as to offer (aka multiprocessing).

Reply:
Yes, definitely the training with all bells and whistles needs to live in the model which it does on the TF side. I'm fine with this, I can show you what I mean in a small refactoring PR afterwards.

Reply:
@DocGarbanzo the point of the [PyTorch Lightning Data Module](https://pytorch-lightning.readthedocs.io/en/latest/datamodules.html) is to encapsulate all the logic regarding data preparation:

```
A datamodule encapsulates the five steps involved in data processing in PyTorch:
1. Download / tokenize / process.
2. Clean and (maybe) save to disk.
3. Load inside Dataset.
4. Apply transforms (rotate, tokenize, etc…).
5. Wrap inside a DataLoader.
```

This makes the DataModule much more re-usable. Your call, but moving the transforms into the model might lose the benefits of having the DataModule. Additionally, a lot of the models will probably be applying similar augmentations, so it might be helpful for these to be coupled with the data vs. with the model.